### PR TITLE
CIDEVSTC-290 Data Product Metadata and Catalog Management

### PR DESCRIPTION
**Pull request by lukecampbell**
- Adopts a template based system to allow a more granular externalization for data product metadata
- Provides individual CRUD operators for catalog entries of data products
- Adds consistency for the catalog by keeping a working copy in the Directory.
- Changes to the catalog are atomic in nature now, so it's less prone to failures because of us adding it in the Directory.
- Considerably more metadata are available in ERDDAP now.

### CODE REVIEW

**Comment by lukecampbell** [comment]
@MauriceManning @mmeisinger Ready for review please.

**Comment by lukecampbell** [comment]
http://buildbot.oceanobservatories.org:8010/builders/coi_any/builds/232

**Comment by lukecampbell** [comment]
Oh yeah forgot, another big change about this is that PyDAP now accesses the container, it's no longer separable (without additional work). It uses the resource registry to identify associated datasets, this was actually in-preparation for M193.

File: ion/services/sa/product/data_product_management_service.py [hunk 1]
```diff
@@ -227,9 +218,21 @@ def update_data_product(self, data_product=None):
...
         self.RR2.update(data_product, RT.DataProduct)
+        if self._metadata_changed(original, data_product):
+            self.update_catalog_entry(data_product._id)
+
+    def _metadata_changed(self, original_dp, new_dp):
+        from ion.processes.data.registration.registration_process import RegistrationProcess
+        for field in RegistrationProcess.catalog_metadata:
+            if hasattr(original_dp, field) and getattr(original_dp, field) != getattr(new_dp, field):
+                return True
+        return False
```
Comment (lukecampbell) [inline ion/services/sa/product/data_product_management_service.py:232]: I wanted to point this out, we'll only update a catalog entry through update_data_product if the metadata fields have changed. I'm trying to reduce the frequency at which we could potentially update the catalog.

**Comment by mmeisinger** [comment]
Looks good to me (not knowing the details). Good use of the directory.

**Comment by lukecampbell** [comment]
http://buildbot.oceanobservatories.org:8010/builders/coi_any/builds/234

**Pull request merged**